### ISSUE TITLE
[sdk] dsl.ExitHandler error: bad replacement for`{{workflow.failures}}` placeholder results in SyntaxError

### ISSUE THREAD

**Issue by jli** [kubeflow/pipelines#5858]
I ran into this issue while trying to use ExitHandler for workflow completion/failure notifications, see https://github.com/kubeflow/pipelines/issues/3516 for context.


### Environment

*  KFP version: 1.4.1
*  KFP SDK version: 1.4.0
*  All dependencies version:
   * kfp                           1.4.0
   * kfp-pipeline-spec             0.1.7
   * kfp-server-api                1.5.0


### Steps to reproduce

- Try to use `dsl.ExitHandler` with a function that uses workflow variable `{{workflow.failures}}`. (See sample pipeline code at the bottom of this issue.)
- Launch pipeline run on KFP.
- Get error `SyntaxError: invalid syntax`.

My exit handler component uses a bunch of other workflow variables, and all the other ones seem to work. When I remove `{{workflow.failures}}`, my exit handler runs correctly.

When I change my example task to fail so that the `workflow.failures` variable is populated, I get a different SyntaxError (see below for full details).


### Materials and Reference

Here's the full pipeline that results in the syntax error:

```python
from kfp import dsl
import kfp.components as comp


@dsl.pipeline(name="Pipeline for testing exit handler")
def exit_handler_test_pipeline():
    @comp.create_component_from_func
    def exit_handler_component(
        name: str = "{{workflow.name}}",
        namespace: str = "{{workflow.namespace}}",
        uid: str = "{{workflow.uid}}",
        creationTimestamp: str = "{{workflow.creationTimestamp}}",
        failures: str = "{{workflow.failures}}",
        status: str = "{{workflow.status}}",
    ):
        print(f"my test exit handler!")
        print(f"name: {name}")
        print(f"namespace: {namespace}")
        print(f"uid: {uid}")
        print(f"creationTimestamp: {creationTimestamp}")
        print(f"failures: {failures}")
        print(f"status: {status}")

    @comp.create_component_from_func
    def some_example_component():
        print("do some task")

    exit_handler_op = exit_handler_component()

    with dsl.ExitHandler(exit_handler_op):
        _example_task = some_example_component()
```

I get the following error in the exit handler task logs on KFP:

```python
  File "/tmp/tmp.jPJXqdBDsc", line 6
    failures = ""null"",
                    ^
SyntaxError: invalid syntax
```


Here's a screenshot of the error from the run on KFP:

![Screenshot of error on KFP](https://user-images.githubusercontent.com/133466/121966574-e04ff380-cd3c-11eb-92ac-2319009f0474.png)



---



**Similar SyntaxError when the pipeline has a task failure**

I also tested the exit handler when there are task failures by changing `some_example_component` to this:

```python
    @comp.create_component_from_func
    def some_example_component():
        print("do some task")
        raise ValueError("cause task failure")
```

This results in `workflow.failures` being populated with some JSON, but this still causes a SyntaxError:

```python
  File "/tmp/tmp.W7x1yGRiSo", line 6
    failures = ""[{\"displayName\":\"some-example-component\",\"message\":\"failed with exit code 1\",\"templateName\":\"some-example-component\",\"phase\":\"Failed\",\"podName\":\"pipeline-for-testing-exit-handler-pcm5g-1464849803\",\"finishedAt\":\"2021-06-14T22:27:44Z\"},{\"displayName\":\"exit-handler-1\",\"message\":\"\",\"templateName\":\"exit-handler-1\",\"phase\":\"Failed\",\"podName\":\"pipeline-for-testing-exit-handler-pcm5g-655720480\",\"finishedAt\":\"2021-06-14T22:27:45Z\"},{\"displayName\":\"pipeline-for-testing-exit-handler-pcm5g\",\"message\":\"\",\"templateName\":\"pipeline-for-testing-exit-handler\",\"phase\":\"Failed\",\"podName\":\"pipeline-for-testing-exit-handler-pcm5g\",\"finishedAt\":\"2021-06-14T22:27:45Z\"}]"",
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         ^
SyntaxError: unexpected character after line continuation character
```

It looks like the underlying problem is the same: KFP (or maybe it's Argo?) is inserting extra double quotes when replacing `"{{workflow.failures}}"`, which is causing a syntax error in Python.

I also tried annotating the `faillures` argument with `Optional[str]` instead of `str`, but the result seems to be the same.


Is there any way to work around this? Is there a way to access these workflow variables via Python code rather than having KFP/Argo replacing these placeholder strings?

---

<!-- Don't delete message below to encourage users to support your issue! -->
Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by jli** [comment]
Hi team, any progress on this issue?

My team is use KFP in production, and we want to add monitoring/alerting to our production pipelines so we know when things fail. The KFP ExitHandler seems like the perfect tool for this, but we cannot use it due to this bug.

Any hints about workarounds would also be appreciated. It seems like it's most likely an underlying Argo problem, right? Will this get fixed with a more recent Argo or KFP release?

**Comment by jli** [comment]
FWIW, I switched to using a yaml-based component definition instead of `create_component_from_func`, and that seems to have fixed the issue with the parse error.

It still seems that the `workflow.failures` is string-escaped twice (as if the value is from `str(json.dumps(failure_object))` instead of just `json_dumps.(failure_object)`, so I had to undo that extra escaping with `ast.literal_eval`.

So, I'm guessing that my original example would have worked if i used single quotes instead of double quotes around `{{workflow.failures}}` in my function definition.
